feat: component macro

davidon-top opened the pull request:
issue: #13 
for now its 1:1 copy of [this](https://github.com/davidon-top/rstml-component-utils-macro) with changed code style
Todo:
- [x] component_html macro to remove formatter variable in write_html while keeping ide helpers
- [ ] maybe: put the util macros behind feature flag
- [x] generate a function instead of completely discarding it
- [x] docs

davidon-top commented:
Implementation is done, last thing that might need to be done is put the utility macros behind a feature flag which i don't know if you want or no.
If yes just let me know and ill add it.

davidon-top commented:
Also i noticed that the `component.rs` test is failing because of a whitespace issue and that in `Cargo.toml` in root uses version 0.2.0 insted of 0.2.1 of rstml-component-macro

Alxandr commented:
I'm hoping to be able to take a look at this tomorrow.

Alxandr commented on macro/src/lib.rs:
````diff
@@ -21,3 +22,28 @@ pub fn write_html(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
 	component::derive_html_component(input.into()).into()
 }
+
+/// makes a function that returns std::fmt::Result into a component function, argument to the
+/// attribute is the name of the struct that will be generated
+/// usage:
+/// ```rust
+/// #[component(MyComponent)]
+/// pub fn my_component(title: String) -> std::fmt::Result {
+///		component_html!(
+///			<div>{title}</div>
+///		)
+/// }
+/// ```
````
```suggestion
/// Makes a function that returns `std::fmt::Result` into a component function, argument to the
/// attribute is the name of the struct that will be generated.
///
/// # Usage
///
/// ```
/// #[component(MyComponent)]
/// pub fn my_component(title: String) -> std::fmt::Result {
///     component_html!(
///         <div>{title}</div>
///     )
/// }
/// ```
```

Alxandr commented on macro/src/lib.rs:
```diff
@@ -21,3 +22,28 @@ pub fn write_html(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+	utils::component(attr.into(), input.into()).into()
+}
+
+/// use inside component function, expands to `write_html!(formatter, #macro_input)`
```
```suggestion
/// Use inside component functions, expands to `write_html!(formatter, #macro_input)`
```

I'm not convinced this macro is needed. I'll write some more about that in the general review comment - but if it's going to be kept, at least the documentation-comment should be formatted properly.

Alxandr commented on macro/src/utils.rs:
```diff
@@ -0,0 +1,150 @@
+		Err(err) => return err.to_compile_error(),
+	};
+	// check if input is valid
+	assert!(
```
Macros should not panic, and should instead output diagnostics. See how `Template` in `template.rs` has a `diagnostics: Vec<Diagnostic>` field. Instead of using asserts, do the check and if the macro should fail, add diagnostic to a vec. Then after all the checks are done, check if the list is empty, and return diagnostics if it's not.

Alxandr commented on macro/src/utils.rs:
```diff
@@ -0,0 +1,150 @@
+	);
+	match input.sig.output {
+		syn::ReturnType::Type(_, t) => {
+			assert_eq!(
```
This is pretty bad. What if I have `use std::fmt::Result as Result;`? You should not expect full paths. Rather than checking the type in the macro, code should be emitted that asserts that the return type is correct.

Alxandr commented on macro/src/utils.rs:
```diff
@@ -0,0 +1,150 @@
+				"component function must return std::fmt::Result"
+			)
+		}
+		_ => panic!("component function must return std::fmt::Result"),
```
Again, don't panic.

Alxandr commented on macro/src/utils.rs:
```diff
@@ -0,0 +1,150 @@
+				let ty = *pat.ty.clone();
+				let ident = match *pat.pat.clone() {
+					syn::Pat::Ident(pat_ident) => pat_ident,
+					_ => panic!("component function must have typed arguments"),
```
No panic. Also I think this error message is wrong.

Alxandr commented on macro/src/utils.rs:
```diff
@@ -0,0 +1,150 @@
+					_ => panic!("component function must have typed arguments"),
+				};
+				arg_vars.push(quote!(let #ident = self.#ident;));
+				struct_args.push(quote!(#visibility #ident: #ty,));
```
This should probably just be `pub`. (as opposed to `#visibility`) It doesn't really matter if the struct fields are pub even if the struct is not.

Alxandr commented on macro/src/utils.rs:
```diff
@@ -0,0 +1,150 @@
+				struct_args.push(quote!(#visibility #ident: #ty,));
+				idents.push(quote!(#ident));
+			}
+			_ => panic!("component function must have typed arguments"),
```
No panics.

Alxandr commented on macro/src/utils.rs:
```diff
@@ -0,0 +1,150 @@
+		}
+	}
+
+	let mut gen_struct: ItemStruct = parse_quote! {
```
I don't think this is a good idea. Don't generate and parse your own code in the macros. Syn has heplers for getting generics correctly.

Alxandr commented on macro/src/utils.rs:
```diff
@@ -0,0 +1,150 @@
+	impl_generics.where_clause = None; // this gets copied to impl so remove it for Struct<>
+
+	let fn_body = input.block;
+	let mut impl_component: ItemImpl = parse_quote! {
```
Don't parse your own generated code.

Alxandr reviewed (commented):
I've reviewed most of the code here - but I'd like to discuss the overall design of the macros a bit before committing to this exact design, as I think we can do it a bit better.

If we take a look at the example used in the code here:

```rust
#[component(MyComponent)]
fn my_component(title: String) -> std::fmt::Result {
	component_html!(
		<div>{title}</div>
	)
}
```

this currently turns into something like this:

```rust
#[derive(Component)]
struct MyComponent {
  title: String,
}

impl HtmlContent for MyComponent {
  fn fmt(self, formatter: &mut HtmlFormatter) -> fmt::Result {
    let title = self.title;
    write_html!(formatter, <div>{title}</div>)
  }
}

fn my_component(title: String) -> MyComponent {
  MyComponent { title }
}
```

Now, I think there are a couple of issues with this. The first one is that the signature of the `my_component` is changed. This is a bit unfortunate I think, as it get's harder to understand the code as written. The second part is that I think the `component_html` macro is largely useless - and can probably be removed by also fixing the first point.

So I instead suggest doing the following:

```rust
#[component(MyComponent)]
// #[component(pub MyComponent)] - maybe we also allow this - for getting the struct-only be pub?
fn my_component(title: String) -> impl HtmlContent {
	html!(
		<div>{title}</div>
	)
}
```

This means you don't actually have to touch the original implementation at all. And the generated code would look like this:

```rust
#[derive(Component)]
struct MyComponent {
  title: String,
}

impl HtmlContent for MyComponent {
  fn fmt(self, formatter: &mut HtmlFormatter) -> fmt::Result {
    formatter.write_content(my_component(self.title))
  }
}
```

At the same time, the `html` macro should probably get move-semantics by default, removing `move_html`. What do you think?

davidon-top commented:
> So I instead suggest doing the following:

So i thought of something similar, but i didn't want to touch the original html! implementation. I waned this to be "an extension" to what you already have (if some1 wants to they should still be able to write components as structs) and so changing the html! implementation to return something else then fmt::Result would change how the struct components are written as well. Maybe there can be 2 macros one for function style components and one for sturct like? or maybe add helpers to make it as simple as it was now to still make struct components with a new html macro.

Thanks for all the tips for writing macros, ill fix and improve what i can.
I only realized now that what i've made was good enough for me, what i should have done (and probably will soon) is make it good for everyone (or most people, nothings perfect)

davidon-top commented:
fyi: force push to change typo in commit message

Alxandr commented:
The `html!` macro today isn't used, nor does it return `std::fmt::Result`. It returns an `impl FnOnce(&mut HtmlFormatter) -> std::fmt::Result` (which in turn impls `HtmlContent`). So you wouldn't have to change the `html!` macro - except for probably giving it `move` semantics, which is literally just changing `false` to `true` in the `lib.rs` file [here](https://github.com/YoloDev/rstml-component/blob/b287039/macro/src/lib.rs#L7C28-L7C33).

davidon-top commented:
I'm kind of busy with school now, this has to be put on hold. Sorry

Alxandr commented:
@fraschm1998 why do you need this PR to start trying out rstml-component?

davidon-top commented:
School and other stuff got more chill so going to continue working on this.

Alxandr commented:
Poke me when you feel like you want a review - or help for that matter.

davidon-top commented:
Going to do (or at least try to do) some optimizations, there are 39 occurrences of .clone()
It should be feature complete tho.

davidon-top commented:
@Alxandr Should be ready for another review. I fixed and implemented everything all the way from diagnostics instead of panicking to argument extractors `fn ...(Foo(title): Foo) ...` and better generics.

Alxandr commented on tests/utils.rs:
```diff
@@ -0,0 +1,50 @@
+}
+
+#[component(MyGenericComponent)]
+fn my_generic_component(title: impl Into<String>) -> impl HtmlContent {
```
Could you add a couple of more generics tests?

I'm thinking:

```rust
fn generics_with_where<T>(title: T) -> impl HtmlContent 
where T: HtmlContent { ... }

fn inline_clause<T: HtmlContent> .......
fn mixed<T: HtmlContent>(title: T, body: impl HtmlContent) ....
```

davidon-top replied:
Good call, there was a bug where for some reason expanding Generics doesn't include the where clause

davidon-top commented:
@Alxandr review?

CC15XX1 reviewed (commented):
lgtm

Alxandr commented:
From the test it seems reasonable - but as you're probably well aware macro code can be pretty tricky to get right, so I'm not comfortable simply reading the code and going "looks good to me". I will try to take some time either tomorrow or the day after to properly understand this.

That isn't to say I think it's wrong, just so you don't get that impression. I simply don't trust myself enough with code like this without properly understanding it. Especially since I might end up maintaining it.

Alxandr reviewed (commented):
I'm really quite happy with the output of the macro now (with a very minor exception that I'd like the original `fn` to be emitted before the generated code).

But before I get to the review details I'd like to take a moment to talk about the personal side of this process, cause I'm not used to doing review processes like this with "random people on the internet". And since I don't know you, I don't know what your expected/preferred outcome is. There's a couple of options we could take from here, and both of them are equally fine for me - so I would like to know what your preferred resolution is before continuing.

1. I could take the code you've submitted here, fix it up to my preferences, and then merge it. It would not change the API of the macro, it's "finished" in my eyes (with the minor exception of what happens if you don't specify the visibility). That way you'd get the macro sooner, and would be able to build things with it.
2. We could continue this back-and-forth where I review the code and tell you what I think needs to be improved. From what I read from this latest version, it seems to me that you're somewhat new to rust (though I might obviously be reading this wrong). I'm quite impressed by how you've dug into macros if that is indeed the case though, cause it's not for everybody. I am also available on more immediate communication platforms (like discord for instance) if you want to have a more in-depth back-and-forth about specific pieces of code, or get more of my reasoning for the comments I leave.

With that out of the way, if you want to continue working on this, I've left comments for the local issues I've found in the code. Beyond what's commented in the code, there are a decent amount of `.clone()` calls that are very unnecessary. While I'm not trying to say "never clone", nor even "never clone unnecessarily", some of these clones are pretty bad - especially if they are used for a recursive function that doesn't actually need it to be cloned, like the `generate_one` function.

Let me know if you have any questions. It doesn't matter if they are small.

davidon-top commented:
> it seems to me that you're somewhat new to rust

im new to proc macro dev, i only made simple macros for internal use, i would otherwise consider myself as "becoming intermediate" (i started using rust at around 1.56 +- a few versions). Also with the .clone spam im usually lazy and just put .clone telling myself "yeh ill fix this later" then forget and never fix it.

I can fix the current issues in the review (because you already told me what you want changed) and then i don't care if you want to take over or no. I don't know if its easier for you to just fix the code or to review the code.

Alxandr commented:
> and then i don't care if you want to take over or no. I don't know if its easier for you to just fix the code or to review the code.

I'm fine with both. Do I believe I could do it faster? Yes. But that's mostly due to the asynchronous nature of this communication. I also don't need this to be done particularly fast, as I'm not using this code anywhere at the moment (I have hardware issues on the system that was supposed to use this library which I need to deal with first). So if it takes a few weeks, that's not an issue to me.

Just let me know if you get tired off working on it and I can take over.

davidon-top commented on macro/src/utils.rs:
```diff
@@ -0,0 +1,371 @@
+	TypeTuple, Visibility,
+};
+
+struct AttrParser {
```
imo: ComponentAttrs is a better name as that as that suggests that its probably the macro attributes

davidon-top commented on macro/src/utils.rs:
```diff
@@ -0,0 +1,371 @@
+			.push(syn::Error::new_spanned(abi, "component function must not have an abi").into());
+	}
+
+	let mut generics = input.sig.generics.clone();
```
I need to use builders/factories more i always forget they exist.

davidon-top commented:
4 more conversations to resolve / implement

davidon-top commented:
Sorry for the commit mess, i have no idea what happened there. I was trying to update my fork with 1 commit from your repo and when i now commited new changes git decided to double most commits? i think. if it bothers you i can try to fix it with somehow changing the git history locally and force push

davidon-top commented:
If you want you can take it from here. Or whatever suits you more. Currently i am not doing webdev where i would need this so I'm not in a hurry.

Alxandr commented:
> Currently i am not doing webdev where i would need this so I'm not in a hurry.

My situation is the same :P. I'll take a look at what's here at least. Don't worry about the commits being wonky. I squash all PR merges.

davidon-top commented:
Any updates? (i don't want to pressure you or be rude. i just want to know if thers any progress on this (or if i should fix some stuff), i understand you're proboably bussy)

Alxandr commented:
Ah. My memory of this PR was that I was waiting for you, so I kinda completely forgot about it :P. I can take over if you want to, but it will probably not be for about a week at least, as next week is kinda hectic. Hopefully I have time to spend on this during the holidays.

Alxandr commented:
I'm going through this code again now (been a while), preparing it for merging - and while doing so I started thinking: does it actually make sense to support things like tuple arguments and such? I mean, the whole point of the function macro is to simplify creating structs with named fields - so that you can construct them in "html". But if I do something like this:

```rust
#[component(Comp)]
fn comp((a, b): (String, String)) -> impl HtmlComponent {
   ....
}
```

I can't really name the tuple. So I will have to do something like `<Comp tuple__1={(a, b)] />`. Is this useful? I don't think so, so I will remove the support for "fancy" arguments for now - and we can rather think about how this should work later.

I will sill allow support for fancy argument types (like `impl` args), they just have to be named.

davidon-top commented:
tuples not so much but extractors eg: Query(q): Query<T> is usefull and i think struct extractors Thing {a, b}: Thing are also usefull, and if you have that you might as well have all the types

Alxandr commented:
Please explain to me how they are useful. I would need a concrete example.

Alxandr commented:
This PR is now ready to merge from my side. I'm waiting to see if you have any good examples for needing parameter bindings first however.

davidon-top commented:
lets say i have a axum handler that is fn(Json(body): Json<SomeStruct>) and SomeStruct has fields foo, bar, x, y. and i have <div><Comp1 c={&body} /><Comp2 c={&body} /></div>.
Then Comp1 would need foo and bar out of the response so it would accept fn(SomeStruct {foo, bar, ..}: SomeStruct) and Comp2 would need bar and x so fn(SomeStruct {bar, x, ..}: SomeStruct).
I was using my fork in a project and i have used struct extrctors in a similar way

Alxandr commented:
You can't name the arguments though, so you end up having to do `<MyComp some_random_prop__1={&body} />`. It's brittle (the name of the prop is order-dependent) and not very discoverable. Also, you can easily get the same doing this:

```rust
#[component]
fn comp_func(body: SomeStruct) -> impl HtmlContent {
  html!(<div>{body.bar}</div>)
}

Alxandr commented:
Anyways, I'm going to merge this now. If complex bindings are shown to be useful, they can easily be added later - but I am not willing to add them before a proper naming strategy for the component struct props is defined.

davidon-top commented:
looking back at my code i did use it but only for top components that get called by axum not in markup

Alxandr commented:
Those aren't components. They are just functions that return `impl HtmlContent`. You don't need the macro for that. And I agree that those are useful, but not with the macro.
